### DISCUSSION TITLE
Training network for multiple camera types?

### DISCUSSION THREAD

**Discussion by shwan19** [talmolab/sleap#849] [Help!]
Hi SLEAP team,

I initially trained a successful SLEAP network on videos with high spatial and temporal resolution. I now hope to generalize this network to videos taken from cameras with lower resolution. These lower-resolution videos are of the same types of animals and have the same types of social behaviors. I quickly realized that my existing network does not perform well on lower-resolution videos, so I labeled 20 frames from each of my lower-resolution videos. But when I try to train SLEAP from baseline bottom-up network using a combination of my 1000+ labeled frames from high-resolution videos and ~200 frames from low-resolution videos, the network stopped training quite early (presumably due to plateaued loss). This hybrid dataset-trained network was not effective for inference on frames from either the high-resolution or lower-resolution videos. (I also recognize that a lot of the network hyperparameters are pixel-based, which would be really difficult to reconcile across videos with very different spatial resolution.)

I could start training a new network for each camera type, but ideally I hope to take advantage of the powerful capabilities of the existing network. I was wondering if you have guidance on the following.

1) Is it realistic to have a single network that performs inference on videos with very different spatial dimensions / amount of graininess?

2) What are best practices for jointly handling videos of different quality?

3) Are ways to transfer abstracted pose representation from the highly effective neural network (trained on frames from high-resolution recordings) to a new set of training frames derived from lower-resolution recordings? I.e. If easily identifiable, freeze a layer that has pose representation?

Many, many thanks in advance for your guidance!

Shuyu

**Comment by talmo** [comment]
Hi @shwan19,

Sorry for the delay! This is a great question.

I think that you're going to have a hard time generalizing across the two conditions. Ultimately, the feature detectors (especially in the early layers of the network) are going to be resolution-specific. Even if we're able impose some scale invariance through scale augmentation, you'll end up sacrificing accuracy in one or the other.

Generally speaking, the philosophy we adopted with SLEAP is that it should be easy and fast to generate new labeled datasets for new conditions and be able to train models with relatively few labels. This "specialist" approach is often more practical than the "generalist" approach of developing models that can handle arbitrary conditions, especially if it sacrifices accuracy in either condition for generality across both.

The idea of leveraging structure in poses (independent of image statistics) is a great one, but unfortunately still very much a research question as to how it could be best leveraged and practically implemented. Separating image-independent features from those in images may not be as easy as finding the right layer without imposing additional constraints -- this is typically called disentanglement in computer vision/deep learning, and is also an active area of research. (BTW: If you're interested in trying it out, we'd love to chat with you!)

For now, I think the most practical way to take advantage of your high resolution dataset to inform your low one would be to downsample the high resolution images to match the low ones before pooling them together. That would at least take care of the scale-invariance issue, and then you're just left with the noise profiles being different (which is easier for networks to deal with).

Technically, the temporal resolution mismatch shouldn't matter since we train on single frames. Where that might come into play is if the differing temporal resolutions also affect other imaging properties such as exposure time (longer means more motion blur in the low FPS images) and shot noise (higher in the high FPS images since the exposure is lower). Ideally you'd have a short exposure for both and identical illumination conditions, but I appreciate that that's not usually the easiest thing to set up. To start I would just not worry about it and see how far you get.

Let me know if you have any questions or need clarifications! And if you do try out the approach above, we'd be very eager to hear how it worked out for you!

Cheers,

Talmo
